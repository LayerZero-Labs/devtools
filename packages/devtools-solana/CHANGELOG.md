# @layerzerolabs/devtools-solana

## 0.1.1

### Patch Changes

- 10dcf5b: Add schemas for working with BN & PublicKey objects

## 0.1.0

### Minor Changes

- 1f18418: Allow EndpointV2.setConfig to return more than one transaction

## 0.0.6

### Patch Changes

- 4d6cf0d: Add the ability to specify lookup tables in OmniSignerSolana

## 0.0.5

### Patch Changes

- 753a9aa: Add getPoint method to OmniSigner
- 0db7580: Add createRpcUrlFactory convenience method
- Updated dependencies [753a9aa]
- Updated dependencies [f22da2e]
  - @layerzerolabs/devtools@0.3.23

## 0.0.4

### Patch Changes

- 783461e: Use Asyncretriable when creating Solana transactions
- Updated dependencies [783461e]
  - @layerzerolabs/devtools@0.3.22

## 0.0.3

### Patch Changes

- 9136039: Add publicKey to Solana OmniSDK

## 0.0.2

### Patch Changes

- d393679: Add a stub of a signer and serde utilities for Solana transactions
