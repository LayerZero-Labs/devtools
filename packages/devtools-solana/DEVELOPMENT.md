@solana-developers/helpers pinned to 2.8.0 due to this issue: https://solana.stackexchange.com/questions/21945/issue-with-solana-developers-helpers-package-in-browser-environments-anchors?utm_source=chatgpt.com