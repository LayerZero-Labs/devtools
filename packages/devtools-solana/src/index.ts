export * from './connection'
export * from './omnigraph'
export * from './transactions'
