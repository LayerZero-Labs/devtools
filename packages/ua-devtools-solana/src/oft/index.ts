export * from './config'
export * from './factory'
export * from './sdk'
