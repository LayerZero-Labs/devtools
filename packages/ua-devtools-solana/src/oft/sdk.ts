import type { IOApp, OAppEnforcedOptionParam } from '@layerzerolabs/ua-devtools'
import { OftTools, EndpointProgram, OftProgram } from '@layerzerolabs/lz-solana-sdk-v2'
import {
    type OmniAddress,
    type OmniTransaction,
    formatEid,
    areBytes32Equal,
    makeBytes32,
    Bytes,
    OmniPoint,
    normalizePeer,
    denormalizePeer,
    fromHex,
    toHex,
} from '@layerzerolabs/devtools'
import type { EndpointId } from '@layerzerolabs/lz-definitions'
import { EndpointV2 } from '@layerzerolabs/protocol-devtools-solana'
import { type Logger, printBoolean, printJson } from '@layerzerolabs/io-devtools'
import { mapError, AsyncRetriable } from '@layerzerolabs/devtools'
import { OmniSDK } from '@layerzerolabs/devtools-solana'
import { Connection, PublicKey, Transaction } from '@solana/web3.js'
import { Options } from '@layerzerolabs/lz-v2-utilities'
import assert from 'assert'

export class OFT extends OmniSDK implements IOApp {
    constructor(
        connection: Connection,
        point: OmniPoint,
        userAccount: PublicKey,
        public readonly programId: PublicKey,
        logger?: Logger
    ) {
        super(connection, point, userAccount, logger)
    }

    @AsyncRetriable()
    async getOwner(): Promise<OmniAddress | undefined> {
        this.logger.debug(`Getting owner`)

        const config = await mapError(
            () => OftProgram.accounts.OftConfig.fromAccountAddress(this.connection, this.publicKey),
            (error) => new Error(`Failed to get owner for ${this.label}: ${error}`)
        )

        const owner = config.admin.toBase58()

        return this.logger.debug(`Got owner: ${owner}`), owner
    }

    async hasOwner(address: OmniAddress): Promise<boolean> {
        this.logger.debug(`Checking whether ${address} is an owner`)

        const owner = await this.getOwner()
        const isOwner = areBytes32Equal(normalizePeer(address, this.point.eid), normalizePeer(owner, this.point.eid))

        return this.logger.debug(`Checked whether ${address} is an owner (${owner}): ${printBoolean(isOwner)}`), isOwner
    }

    async setOwner(address: OmniAddress): Promise<OmniTransaction> {
        this.logger.debug(`Setting owner to ${address}`)

        const transaction = await mapError(
            async () =>
                new Transaction().add(
                    await OftTools.createTransferAdminIx(
                        this.programId,
                        this.userAccount, // Signer
                        this.publicKey, // OFT Config account
                        new PublicKey(address) // Owner account
                    )
                ),
            (error) => new Error(`Failed to set owner for ${this.label} to ${address}: ${error}`)
        )

        return {
            ...(await this.createTransaction(transaction)),
            description: `Setting owner to ${address}`,
        }
    }

    @AsyncRetriable()
    async getEndpointSDK(): Promise<EndpointV2> {
        this.logger.debug(`Getting EndpointV2 SDK`)

        return new EndpointV2(
            this.connection,
            { eid: this.point.eid, address: EndpointProgram.PROGRAM_ID.toBase58() },
            this.userAccount
        )
    }

    @AsyncRetriable()
    async getPeer(eid: EndpointId): Promise<OmniAddress | undefined> {
        const eidLabel = `eid ${eid} (${formatEid(eid)})`

        this.logger.debug(`Getting peer for ${eidLabel}`)
        try {
            const peer = await OftTools.getPeerAddress(this.connection, this.programId, this.publicKey, eid)

            // We run the hex string we got through a normalization/denormalization process
            // that will ensure that zero addresses will get stripped
            // and any network-specific logic will be applied
            return denormalizePeer(fromHex(peer), eid)
        } catch (error) {
            if (String(error).match(/Unable to find Peer account at/i)) {
                return undefined
            }

            throw new Error(`Failed to get peer for ${eidLabel} for OFT ${this.label}: ${error}`)
        }
    }

    async hasPeer(eid: EndpointId, address: OmniAddress | null | undefined): Promise<boolean> {
        const peer = await this.getPeer(eid)

        return areBytes32Equal(normalizePeer(peer, eid), normalizePeer(address, eid))
    }

    async setPeer(eid: EndpointId, address: OmniAddress | null | undefined): Promise<OmniTransaction> {
        const eidLabel = formatEid(eid)
        // We use the `mapError` and pretend `normalizePeer` is async to avoid having a let and a try/catch block
        const normalizedPeer = await mapError(
            async () => normalizePeer(address, eid),
            (error) =>
                new Error(`Failed to convert peer ${address} for ${eidLabel} for ${this.label} to bytes: ${error}`)
        )
        const peerAsBytes32 = makeBytes32(normalizedPeer)

        this.logger.debug(`Setting peer for eid ${eid} (${eidLabel}) to address ${peerAsBytes32}`)

        const transaction = new Transaction().add(
            await OftTools.createSetPeerIx(this.programId, this.userAccount, this.publicKey, eid, normalizedPeer)
        )

        return {
            ...(await this.createTransaction(transaction)),
            description: `Setting peer for eid ${eid} (${eidLabel}) to address ${peerAsBytes32}`,
        }
    }

    @AsyncRetriable()
    async getDelegate(): Promise<OmniAddress | undefined> {
        this.logger.debug(`Getting delegate`)

        const endpointSdk = await this.getEndpointSDK()
        const delegate = await endpointSdk.getDelegate(this.point.address)

        return this.logger.verbose(`Got delegate: ${delegate}`), delegate
    }

    @AsyncRetriable()
    async isDelegate(delegate: OmniAddress): Promise<boolean> {
        this.logger.debug(`Checking whether ${delegate} is a delegate`)

        const endpointSdk = await this.getEndpointSDK()
        const isDelegate = await endpointSdk.isDelegate(this.point.address, delegate)

        return this.logger.verbose(`Checked delegate: ${delegate}: ${printBoolean(isDelegate)}`), isDelegate
    }

    async setDelegate(delegate: OmniAddress): Promise<OmniTransaction> {
        this.logger.debug(`Setting delegate to ${delegate}`)

        const transaction = await mapError(
            async () => {
                const instruction = await OftTools.createSetDelegateIx(
                    this.programId,
                    this.userAccount,
                    this.publicKey,
                    new PublicKey(delegate)
                )
                return new Transaction().add(instruction)
            },
            (error) => new Error(`Failed to set delegate for ${this.label} to ${delegate}: ${error}`)
        )

        return {
            ...(await this.createTransaction(transaction)),
            description: `Setting delegate to ${delegate}`,
        }
    }

    @AsyncRetriable()
    async getEnforcedOptions(eid: EndpointId, msgType: number): Promise<Bytes> {
        // First we check that we can understand the message type
        this.assertMsgType(msgType)

        const eidLabel = `eid ${eid} (${formatEid(eid)})`
        this.logger.verbose(`Getting enforced options for ${eidLabel} and message type ${msgType}`)

        try {
            const options = await OftTools.getEnforcedOptions(this.connection, this.programId, this.publicKey, eid)
            const optionsForMsgType = msgType === MSG_TYPE_SEND ? options.send : options.sendAndCall

            return toHex(optionsForMsgType)
        } catch (error) {
            if (String(error).match(/Unable to find EnforcedOptions account/)) {
                return toHex(new Uint8Array(0))
            }

            throw new Error(
                `Failed to get enforced options for ${this.label} for ${eidLabel} and message type ${msgType}: ${error}`
            )
        }
    }

    async setEnforcedOptions(enforcedOptions: OAppEnforcedOptionParam[]): Promise<OmniTransaction> {
        this.logger.verbose(`Setting enforced options to ${printJson(enforcedOptions)}`)

        const transaction = new Transaction()
        const optionsByEidAndMsgType = this.reduceEnforcedOptions(enforcedOptions)
        const emptyOptions = Options.newOptions().toBytes()

        for (const [eid, optionsByMsgType] of optionsByEidAndMsgType) {
            const sendOption = optionsByMsgType.get(MSG_TYPE_SEND) ?? emptyOptions
            const sendAndCallOption = optionsByMsgType.get(MSG_TYPE_SEND_AND_CALL) ?? emptyOptions

            const instruction = await OftTools.createSetEnforcedOptionsIx(
                this.programId, // OFT Program ID
                this.userAccount, // your admin address
                this.publicKey, // your OFT Config
                eid, // destination endpoint id for the options to apply to
                sendOption,
                sendAndCallOption
            )

            transaction.add(instruction)
        }

        return {
            ...(await this.createTransaction(transaction)),
            description: `Setting enforced options to ${printJson(enforcedOptions)}`,
        }
    }

    async isNonceInitialized(eid: EndpointId, peer: OmniAddress): Promise<boolean> {
        const endpointSdk = await this.getEndpointSDK()
        return endpointSdk.isOAppNonceInitialized(this.point.address, eid, peer)
    }

    async initializeNonce(eid: EndpointId, peer: OmniAddress): Promise<[OmniTransaction] | []> {
        this.logger.verbose(`Initializing OApp nonce for peer ${peer} on ${formatEid(eid)}`)

        const endpointSdk = await this.getEndpointSDK()
        return endpointSdk.initializeOAppNonce(this.point.address, eid, peer)
    }

    async isSendLibraryInitialized(eid: EndpointId): Promise<boolean> {
        const endpointSdk = await this.getEndpointSDK()
        return endpointSdk.isSendLibraryInitialized(this.point.address, eid)
    }

    async initializeSendLibrary(eid: EndpointId): Promise<[OmniTransaction] | []> {
        this.logger.verbose(`Initializing send library on ${formatEid(eid)}`)

        const endpointSdk = await this.getEndpointSDK()
        return endpointSdk.initializeSendLibrary(this.point.address, eid)
    }

    async isReceiveLibraryInitialized(eid: EndpointId): Promise<boolean> {
        const endpointSdk = await this.getEndpointSDK()
        return endpointSdk.isReceiveLibraryInitialized(this.point.address, eid)
    }

    async initializeReceiveLibrary(eid: EndpointId): Promise<[OmniTransaction] | []> {
        this.logger.verbose(`Initializing receive library on ${formatEid(eid)}`)

        const endpointSdk = await this.getEndpointSDK()
        return endpointSdk.initializeReceiveLibrary(this.point.address, eid)
    }

    async isOAppConfigInitialized(eid: EndpointId): Promise<boolean> {
        const endpointSdk = await this.getEndpointSDK()
        return endpointSdk.isOAppConfigInitialized(this.point.address, eid)
    }

    async initializeOAppConfig(eid: EndpointId, lib: OmniAddress | null | undefined): Promise<[OmniTransaction] | []> {
        this.logger.verbose(`Initializing OApp config for library ${lib} on ${formatEid(eid)}`)

        const endpointSdk = await this.getEndpointSDK()
        return endpointSdk.initializeOAppConfig(this.point.address, eid, lib ?? undefined)
    }

    /**
     * Helper utility that takes an array of `OAppEnforcedOptionParam` objects and turns them into
     * a map keyed by `EndpointId` that contains another map keyed by `MsgType`.
     *
     * @param {OAppEnforcedOptionParam[]} enforcedOptions
     * @returns {Map<EndpointId, Map<MsgType, Uint8Array>>}
     */
    private reduceEnforcedOptions(
        enforcedOptions: OAppEnforcedOptionParam[]
    ): Map<EndpointId, Map<MsgType, Uint8Array>> {
        return enforcedOptions.reduce((optionsByEid, enforcedOption) => {
            const {
                eid,
                option: { msgType, options },
            } = enforcedOption

            // First we check that we can understand the message type
            this.assertMsgType(msgType)

            // Then we warn the user if they are trying to specify enforced options for eid & msgType more than once
            // in which case the former option will be ignored
            const optionsByMsgType = optionsByEid.get(eid) ?? new Map<MsgType, Uint8Array>()
            if (optionsByMsgType.has(msgType)) {
                this.logger.warn(`Duplicate enforced option for ${formatEid(eid)} and msgType ${msgType}`)
            }

            // We wrap the call with try/catch to deliver a better error message in case malformed options were passed
            try {
                optionsByMsgType.set(msgType, Options.fromOptions(options).toBytes())
            } catch (error) {
                throw new Error(
                    `Invalid enforced options for ${this.label} for ${formatEid(eid)} and msgType ${msgType}: ${options}: ${error}`
                )
            }

            optionsByEid.set(eid, optionsByMsgType)

            return optionsByEid
        }, new Map<EndpointId, Map<MsgType, Uint8Array>>())
    }

    /**
     * Helper method that asserts that `value` is a `MsgType` that the OFT understands
     * and prints out a friendly error message if it doesn't
     *
     * @param {unknown} value
     * @returns {undefined}
     */
    private assertMsgType(value: unknown): asserts value is MsgType {
        assert(
            isMsgType(value),
            `${this.label}: Invalid msgType received: ${value}. Expected one of ${MSG_TYPE_SEND} (send), ${MSG_TYPE_SEND_AND_CALL} (send and call)`
        )
    }

    async setCallerBpsCap(callerBpsCap: bigint): Promise<OmniTransaction | undefined> {
        this.logger.debug(`Setting caller BPS cap to ${callerBpsCap}`)

        throw new TypeError(`setCallerBpsCap() not implemented on Solana OFT SDK`)
    }

    @AsyncRetriable()
    async getCallerBpsCap(): Promise<bigint | undefined> {
        this.logger.debug(`Getting caller BPS cap`)

        throw new TypeError(`getCallerBpsCap() not implemented on Solana OFT SDK`)
    }
}

type MsgType = 1 | 2

const MSG_TYPE_SEND = 1 satisfies MsgType
const MSG_TYPE_SEND_AND_CALL = 2 satisfies MsgType

const isMsgType = (value: unknown): value is MsgType => value === MSG_TYPE_SEND || value === MSG_TYPE_SEND_AND_CALL
