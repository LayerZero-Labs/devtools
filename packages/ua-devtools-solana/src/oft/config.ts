import {
    type OmniVector,
    type CreateTransactionsFromOmniEdges,
    formatOmniVector,
    createConfigureEdges,
    createConfigureMultiple,
} from '@layerzerolabs/devtools'
import { createModuleLogger } from '@layerzerolabs/io-devtools'
import { endpointIdToChainType, ChainType } from '@layerzerolabs/lz-definitions'
import type { IOApp, OAppConfigurator, OAppOmniGraph } from '@layerzerolabs/ua-devtools'
import { OFT } from './sdk'

const createOFTLogger = () => createModuleLogger('OFT')

/**
 * Helper function that checks whether a vector originates from a Solana network
 *
 * @param {OmniVector} vector
 * @returns {boolean}
 */
const isVectorFromSolana = (vector: OmniVector): boolean => endpointIdToChainType(vector.from.eid) === ChainType.SOLANA

/**
 * Helper function that wraps a edge configuration function,
 * only executing it for edges that originate in Solana
 *
 * @param {CreateTransactionsFromOmniEdges<OAppOmniGraph, IOApp>} createTransactions
 * @returns {CreateTransactionsFromOmniEdges<OAppOmniGraph, IOApp>}
 */
const onlyEdgesFromSolana = (
    createTransactions: CreateTransactionsFromOmniEdges<OAppOmniGraph, IOApp>
): CreateTransactionsFromOmniEdges<OAppOmniGraph, IOApp> => {
    const logger = createOFTLogger()

    return (edge, sdk, graph, createSdk) => {
        if (!isVectorFromSolana(edge.vector)) {
            return logger.verbose(`Ignoring connection ${formatOmniVector(edge.vector)}`), undefined
        }

        return createTransactions(edge, sdk, graph, createSdk)
    }
}

export const initNonce: OAppConfigurator = createConfigureEdges(
    onlyEdgesFromSolana(({ vector }, sdk) => {
        const logger = createOFTLogger()

        if (typeof (sdk as OFT).initializeNonce !== 'function') {
            return logger.warn(`Could not find initializeNonce() method on OFT SDK, skipping`), undefined
        }

        return (sdk as OFT).initializeNonce(vector.to.eid, vector.to.address)
    })
)

export const initSendLibrary: OAppConfigurator = createConfigureEdges(
    onlyEdgesFromSolana(({ vector }, sdk) => {
        const logger = createOFTLogger()

        if (typeof (sdk as OFT).initializeSendLibrary !== 'function') {
            return logger.warn(`Could not find initializeSendLibrary() method on OFT SDK, skipping`), undefined
        }

        return (sdk as OFT).initializeSendLibrary(vector.to.eid)
    })
)

export const initReceiveLibrary: OAppConfigurator = createConfigureEdges(
    onlyEdgesFromSolana(({ vector }, sdk) => {
        const logger = createOFTLogger()

        if (typeof (sdk as OFT).initializeReceiveLibrary !== 'function') {
            return logger.warn(`Could not find initializeReceiveLibrary() method on OFT SDK, skipping`), undefined
        }

        return (sdk as OFT).initializeReceiveLibrary(vector.to.eid)
    })
)

export const initOFTConfig: OAppConfigurator = createConfigureEdges(
    onlyEdgesFromSolana(({ vector, config }, sdk) => {
        const logger = createOFTLogger()

        if (typeof (sdk as OFT).initializeOAppConfig !== 'function') {
            return logger.warn(`Could not find initializeOAppConfig() method on OFT SDK, skipping`), undefined
        }

        return (sdk as OFT).initializeOAppConfig(vector.to.eid, config?.sendLibrary)
    })
)

export const initOFTAccounts = createConfigureMultiple(initNonce, initSendLibrary, initReceiveLibrary, initOFTConfig)
