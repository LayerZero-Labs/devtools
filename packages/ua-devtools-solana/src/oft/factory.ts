import pMemoize from 'p-memoize'
import type { OAppFactory } from '@layerzerolabs/ua-devtools'
import { OFT } from './sdk'
import {
    type ConnectionFactory,
    type PublicKeyFactory,
    createConnectionFactory,
    defaultRpcUrlFactory,
} from '@layerzerolabs/devtools-solana'

/**
 * Syntactic sugar that creates an instance of Solana `OFT` SDK
 * based on an `OmniPoint` with help of an `ConnectionFactory`
 * and a `PublicKeyFactory`
 *
 * @param {PublicKeyFactory} userAccountFactory A function that accepts an `OmniPoint` representing an OFT and returns the user wallet public key
 * @param {PublicKeyFactory} mintAccountFactory A function that accepts an `OmniPoint` representing an OFT and returns the mint public key
 * @param {ConnectionFactory} connectionFactory A function that returns a `Connection` based on an `EndpointId`
 * @returns {OAppFactory<OFT>}
 */
export const createOFTFactory = (
    userAccountFactory: PublicKeyFactory,
    programIdFactory: PublicKeyFactory,
    connectionFactory: ConnectionFactory = createConnectionFactory(defaultRpcUrlFactory)
): OAppFactory<OFT> =>
    pMemoize(
        async (point) =>
            new OFT(
                await connectionFactory(point.eid),
                point,
                await userAccountFactory(point),
                await programIdFactory(point)
            )
    )
