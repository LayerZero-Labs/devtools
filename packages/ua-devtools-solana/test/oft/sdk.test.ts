import { PublicKey } from '@solana/web3.js'
import { createConnectionFactory, defaultRpcUrlFactory } from '@layerzerolabs/devtools-solana'
import { EndpointId } from '@layerzerolabs/lz-definitions'
import { OFT } from '@/oft'
import { makeBytes32, normalizePeer } from '@layerzerolabs/devtools'
import { Options } from '@layerzerolabs/lz-v2-utilities'
import { printJson } from '@layerzerolabs/io-devtools'
import { EndpointProgram, OftTools } from '@layerzerolabs/lz-solana-sdk-v2'

const createSetEnforcedOptionsIxMock = OftTools.createSetEnforcedOptionsIx as jest.Mock

jest.mock('@layerzerolabs/lz-solana-sdk-v2', () => {
    const actual = jest.requireActual('@layerzerolabs/lz-solana-sdk-v2')

    return {
        ...actual,
        OftTools: {
            ...actual.OftTools,
            createSetEnforcedOptionsIx: jest.fn().mockImplementation(actual.OftTools.createSetEnforcedOptionsIx),
        },
    }
})

describe('oft/sdk', () => {
    // FIXME These tests are using a mainnet OFT deployment and are potentially very fragile
    //
    // We need to run our own Solana node with the OFT account cloned
    // so that we can isolate these tests
    const programId = new PublicKey('Ag28jYmND83RnwcSFq2vwWxThSya55etjWJwubd8tRXs')
    const point = { eid: EndpointId.SOLANA_V2_MAINNET, address: '8aFeCEhGLwbWHWiiezLAKanfD5Cn3BW3nP6PZ54K9LYC' }
    const account = new PublicKey('6tzUZqC33igPgP7YyDnUxQg6eupMmZGRGKdVAksgRzvk')

    afterEach(() => {
        createSetEnforcedOptionsIxMock.mockClear()
    })

    describe('getPeer', () => {
        it('should return undefined if we are asking for a peer that has not been set', async () => {
            const connectionFactory = createConnectionFactory(defaultRpcUrlFactory)

            const connection = await connectionFactory(EndpointId.SOLANA_V2_MAINNET)
            const sdk = new OFT(connection, point, account, programId)

            expect(await sdk.getPeer(EndpointId.ETHEREUM_V2_TESTNET)).toBeUndefined()
        })

        it('should return a Solana address if we are asking for a peer that has been set', async () => {
            const connectionFactory = createConnectionFactory(defaultRpcUrlFactory)

            const connection = await connectionFactory(EndpointId.SOLANA_V2_MAINNET)
            const sdk = new OFT(connection, point, account, programId)

            const peer = await sdk.getPeer(EndpointId.ETHEREUM_V2_MAINNET)
            expect(peer).toEqual(expect.any(String))
            expect(normalizePeer(peer, EndpointId.ETHEREUM_V2_MAINNET)).toEqual(expect.any(Uint8Array))

            expect(await sdk.hasPeer(EndpointId.ETHEREUM_V2_MAINNET, peer)).toBeTruthy()
            expect(await sdk.hasPeer(EndpointId.ETHEREUM_V2_MAINNET, undefined)).toBeFalsy()
            expect(await sdk.hasPeer(EndpointId.ETHEREUM_V2_MAINNET, null)).toBeFalsy()
            expect(await sdk.hasPeer(EndpointId.ETHEREUM_V2_MAINNET, makeBytes32())).toBeFalsy()
        })
    })

    describe('setPeer', () => {
        describe('for EVM', () => {
            it('should return an OmniTransaction', async () => {
                const connectionFactory = createConnectionFactory(defaultRpcUrlFactory)

                const connection = await connectionFactory(EndpointId.SOLANA_V2_MAINNET)
                const sdk = new OFT(connection, point, account, programId)

                const omniTransaction = await sdk.setPeer(EndpointId.ETHEREUM_V2_MAINNET, makeBytes32())
                expect(omniTransaction).toEqual({
                    data: expect.any(String),
                    point,
                    description: `Setting peer for eid ${EndpointId.ETHEREUM_V2_MAINNET} (ETHEREUM_V2_MAINNET) to address 0x0000000000000000000000000000000000000000000000000000000000000000`,
                })
            })
        })

        describe('for Aptos', () => {
            it('should return an OmniTransaction', async () => {
                const connectionFactory = createConnectionFactory(defaultRpcUrlFactory)

                const connection = await connectionFactory(EndpointId.SOLANA_V2_MAINNET)
                const sdk = new OFT(connection, point, account, programId)

                const omniTransaction = await sdk.setPeer(EndpointId.APTOS_MAINNET, makeBytes32())
                expect(omniTransaction).toEqual({
                    data: expect.any(String),
                    point,
                    description: `Setting peer for eid ${EndpointId.APTOS_MAINNET} (APTOS_MAINNET) to address 0x0000000000000000000000000000000000000000000000000000000000000000`,
                })
            })
        })

        describe('for Solana', () => {
            it('should return an OmniTransaction', async () => {
                const connectionFactory = createConnectionFactory(defaultRpcUrlFactory)

                const connection = await connectionFactory(EndpointId.SOLANA_V2_MAINNET)
                const sdk = new OFT(connection, point, account, programId)

                const omniTransaction = await sdk.setPeer(EndpointId.SOLANA_V2_MAINNET, point.address)
                expect(omniTransaction).toEqual({
                    data: expect.any(String),
                    point,
                    description: `Setting peer for eid ${EndpointId.SOLANA_V2_MAINNET} (SOLANA_V2_MAINNET) to address ${makeBytes32(normalizePeer(point.address, EndpointId.SOLANA_V2_MAINNET))}`,
                })
            })
        })
    })

    describe('getEnforcedOptions', () => {
        it('should throw if we are trying to get an option for invalid msgType', async () => {
            const connectionFactory = createConnectionFactory(defaultRpcUrlFactory)

            const connection = await connectionFactory(EndpointId.SOLANA_V2_MAINNET)
            const sdk = new OFT(connection, point, account, programId)

            await expect(sdk.getEnforcedOptions(EndpointId.ETHEREUM_V2_TESTNET, 3)).rejects.toMatchSnapshot()
        })

        it('should return an empty bytes if we are asking for options that have not been set', async () => {
            const connectionFactory = createConnectionFactory(defaultRpcUrlFactory)

            const connection = await connectionFactory(EndpointId.SOLANA_V2_MAINNET)
            const sdk = new OFT(connection, point, account, programId)

            expect(await sdk.getEnforcedOptions(EndpointId.ETHEREUM_V2_TESTNET, 1)).toBe('0x')
        })

        it('should return an a hex string if we are asking for options that have been set', async () => {
            const connectionFactory = createConnectionFactory(defaultRpcUrlFactory)

            const connection = await connectionFactory(EndpointId.SOLANA_V2_MAINNET)
            const sdk = new OFT(connection, point, account, programId)

            const sendOptionsHex = await sdk.getEnforcedOptions(EndpointId.ETHEREUM_V2_MAINNET, 1)
            expect(sendOptionsHex).toEqual(expect.any(String))

            const sendOptions = Options.fromOptions(sendOptionsHex)
            expect(sendOptions).toMatchSnapshot()

            const sendAndCallOptionsHex = await sdk.getEnforcedOptions(EndpointId.ETHEREUM_V2_MAINNET, 2)
            expect(sendAndCallOptionsHex).toEqual(expect.any(String))

            const sendAndCallOptions = Options.fromOptions(sendAndCallOptionsHex)
            expect(sendAndCallOptions).toMatchSnapshot()
        })
    })

    describe('setEnforcedOptions', () => {
        it('should throw if we are trying to get an option for invalid msgType', async () => {
            const connectionFactory = createConnectionFactory(defaultRpcUrlFactory)

            const connection = await connectionFactory(EndpointId.SOLANA_V2_MAINNET)
            const sdk = new OFT(connection, point, account, programId)

            await expect(
                sdk.setEnforcedOptions([
                    {
                        eid: EndpointId.ETHEREUM_V2_TESTNET,
                        option: { msgType: 7, options: Options.newOptions().toHex() },
                    },
                ])
            ).rejects.toMatchSnapshot()
        })

        it('should return an omnitransaction if called with valid msgType', async () => {
            const connectionFactory = createConnectionFactory(defaultRpcUrlFactory)

            const connection = await connectionFactory(EndpointId.SOLANA_V2_MAINNET)
            const sdk = new OFT(connection, point, account, programId)

            const enforcedOptions = [
                {
                    eid: EndpointId.ETHEREUM_V2_TESTNET,
                    option: { msgType: 1, options: Options.newOptions().toHex() },
                },
            ]
            const omniTransaction = await sdk.setEnforcedOptions(enforcedOptions)

            expect(omniTransaction).toEqual({
                data: expect.any(String),
                point,
                description: `Setting enforced options to ${printJson(enforcedOptions)}`,
            })
        })

        it('should create instructions grouped by eid', async () => {
            const connectionFactory = createConnectionFactory(defaultRpcUrlFactory)

            const connection = await connectionFactory(EndpointId.SOLANA_V2_MAINNET)
            const sdk = new OFT(connection, point, account, programId)

            const enforcedOptions = [
                // We'll set two options for ethereum
                {
                    eid: EndpointId.ETHEREUM_V2_TESTNET,
                    option: { msgType: 1, options: Options.newOptions().addExecutorOrderedExecutionOption().toHex() },
                },
                {
                    eid: EndpointId.ETHEREUM_V2_TESTNET,
                    option: { msgType: 2, options: Options.newOptions().addExecutorLzReceiveOption(1, 1).toHex() },
                },
                // One option for base
                {
                    eid: EndpointId.BASE_V2_MAINNET,
                    option: { msgType: 1, options: Options.newOptions().addExecutorLzReceiveOption(4, 1).toHex() },
                },
                // And we add a duplicte option for Avalanche
                //
                // The first one should be ignored
                {
                    eid: EndpointId.AVALANCHE_V2_MAINNET,
                    option: { msgType: 2, options: Options.newOptions().addExecutorLzReceiveOption(2, 1).toHex() },
                },
                {
                    eid: EndpointId.AVALANCHE_V2_MAINNET,
                    option: { msgType: 2, options: Options.newOptions().addExecutorLzReceiveOption(3, 1).toHex() },
                },
            ]

            await sdk.setEnforcedOptions(enforcedOptions)

            expect(createSetEnforcedOptionsIxMock).toHaveBeenCalledTimes(3)

            // Ethereum should have both msgType options set
            expect(createSetEnforcedOptionsIxMock).toHaveBeenCalledWith(
                sdk.userAccount,
                sdk.publicKey,
                EndpointId.ETHEREUM_V2_TESTNET,
                Options.newOptions().addExecutorOrderedExecutionOption().toBytes(),
                Options.newOptions().addExecutorLzReceiveOption(1, 1).toBytes(),
                programId
            )

            // Base should have one option set
            expect(createSetEnforcedOptionsIxMock).toHaveBeenCalledWith(
                sdk.userAccount,
                sdk.publicKey,
                EndpointId.BASE_V2_MAINNET,
                Options.newOptions().addExecutorLzReceiveOption(4, 1).toBytes(),
                Options.newOptions().toBytes(),
                programId
            )

            // Avalanche should use the latter option and ignore the first one
            expect(createSetEnforcedOptionsIxMock).toHaveBeenCalledWith(
                sdk.userAccount,
                sdk.publicKey,
                EndpointId.AVALANCHE_V2_MAINNET,
                Options.newOptions().toBytes(),
                Options.newOptions().addExecutorLzReceiveOption(3, 1).toBytes(),
                programId
            )
        })
    })

    describe('getEndpointSDK', () => {
        it('should return an SDK with the correct eid and address', async () => {
            const connectionFactory = createConnectionFactory(defaultRpcUrlFactory)
            const connection = await connectionFactory(EndpointId.SOLANA_V2_MAINNET)
            const sdk = new OFT(connection, point, account, programId)
            const endpointSdk = await sdk.getEndpointSDK()

            expect(endpointSdk.point).toEqual({ eid: sdk.point.eid, address: EndpointProgram.PROGRAM_ID.toBase58() })

            // Run a random function on the SDK to check whether it works
            expect(
                await endpointSdk.isDefaultSendLibrary(sdk.publicKey.toBase58(), EndpointId.ETHEREUM_V2_MAINNET)
            ).toBeFalsy()
        })
    })
})
