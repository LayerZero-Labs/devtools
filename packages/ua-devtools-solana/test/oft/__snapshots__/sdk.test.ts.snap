// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`oft/sdk getEnforcedOptions should return an a hex string if we are asking for options that have been set 1`] = `
_Options {
  "workerOptions": [
    {
      "options": [
        {
          "params": "0x0000000000000000000000000000fde8",
          "type": 1,
        },
      ],
      "workerId": 1,
    },
  ],
}
`;

exports[`oft/sdk getEnforcedOptions should return an a hex string if we are asking for options that have been set 2`] = `
_Options {
  "workerOptions": [
    {
      "options": [
        {
          "params": "0x0000000000000000000000000000fde8",
          "type": 1,
        },
        {
          "params": "0x00000000000000000000000000000000c350",
          "type": 3,
        },
      ],
      "workerId": 1,
    },
  ],
}
`;

exports[`oft/sdk getEnforcedOptions should throw if we are trying to get an option for invalid msgType 1`] = `[AssertionError: Solana program @ [Ag28jYmND83RnwcSFq2vwWxThSya55etjWJwubd8tRXs @ SOLANA_V2_MAINNET]: Invalid msgType received: 3. Expected one of 1 (send), 2 (send and call)]`;

exports[`oft/sdk setEnforcedOptions should throw if we are trying to get an option for invalid msgType 1`] = `[AssertionError: Solana program @ [Ag28jYmND83RnwcSFq2vwWxThSya55etjWJwubd8tRXs @ SOLANA_V2_MAINNET]: Invalid msgType received: 7. Expected one of 1 (send), 2 (send and call)]`;
